benchmarks/bench_functions.cpp

```diff
@@ -0,0 +1,267 @@
+template <Cplx cplx, typename R>
+struct complex_function<cplx, R, FunctionName::SINH> {
+  using T = complex_t<cplx, R>;
+
+  T operator()(const T &a) const {
+    if constexpr (cplx == Cplx::EXT) {
+      return sycl::ext::cplx::sinh(a);
+    } else {
+      return std::sinh(a);
+    }
+  }
+};
+
+template <Cplx cplx, typename R>
+struct complex_function<cplx, R, FunctionName::LOG> {
+  using T = complex_t<cplx, R>;
+
+  T operator()(const T &a) const {
+    if constexpr (cplx == Cplx::EXT) {
+      return sycl::ext::cplx::log(a);
+    } else {
+      return std::log(a);
+    }
+  }
+};
+
+template <Cplx cplx, typename R>
+struct complex_function<cplx, R, FunctionName::SQRT> {
+  using T = complex_t<cplx, R>;
+
+  T operator()(const T &a) const {
+    if constexpr (cplx == Cplx::EXT) {
+      return sycl::ext::cplx::sqrt(a);
+    } else {
+      return std::sqrt(a);
+    }
+  }
+};
+
+template <Cplx cplx, typename R>
+struct complex_function<cplx, R, FunctionName::EXP> {
+  using T = complex_t<cplx, R>;
+
+  T operator()(const T &a) const {
+    if constexpr (cplx == Cplx::EXT) {
+      return sycl::ext::cplx::exp(a);
+    } else {
+      return std::exp(a);
+    }
+  }
+};
+
+template <typename R> class BenchmarkData {
+public:
+  BenchmarkData(std::size_t max_n)
+      : q_(sycl::default_selector_v), max_n_(max_n) {
+    R *h_random_data = sycl::malloc_host<R>(max_n * 4, q_);
+
+    d_a_ = sycl::malloc_device<R>(max_n * 2, q_);
+    d_b_ = sycl::malloc_device<R>(max_n * 2, q_);
+
+    fill_random(h_random_data, max_n * 4);
+    q_.copy(h_random_data, d_a_, 2 * max_n);
+    q_.copy(h_random_data + 2 * max_n, d_b_, 2 * max_n);
+    q_.wait();
+
+    sycl::free(h_random_data, q_);
+  }
+
+  ~BenchmarkData() {
+    sycl::free(d_a_, q_);
+    sycl::free(d_b_, q_);
+  }
+
+  template <typename T> T *get_device_input(std::size_t n) {
+    assert(n <= max_n_);
+    return reinterpret_cast<T *>(d_a_);
+  }
+
+  template <typename T> T *get_device_output(std::size_t n) {
+    assert(n <= max_n_);
+    return reinterpret_cast<T *>(d_b_);
+  }
+
+  sycl::queue &get_queue() { return q_; }
+
+private:
+  sycl::queue q_;
+  std::size_t max_n_;
+  R *h_random_data_;
+  R *d_a_;
+  R *d_b_;
+};
+
+template <typename R> auto get_benchmark_data(std::size_t max_n) {
+  static BenchmarkData<R> *data = nullptr;
+  if (data == nullptr) {
+    data = new BenchmarkData<R>(max_n);
+  }
+  return data;
+}
+
+template <Cplx cplx, typename R, FunctionName F, std::uint32_t SEED = 777>
+static void BM_function(benchmark::State &state) {
+  using T = complex_t<cplx, R>;
+
+  int n = state.range(0);
+
+  complex_function<cplx, R, F> fn{};
+
+  auto bench_data = get_benchmark_data<R>(n);
+
+  auto a = bench_data->template get_device_input<T>(n);
+  auto b = bench_data->template get_device_output<T>(n);
+
+  sycl::queue &Q = bench_data->get_queue();
+
+  for (auto _ : state) {
+    Q.parallel_for(sycl::range<1>(n), [=](sycl::id<1> i) { b[i] = fn(a[i]); });
+    Q.wait();
+  }
```

Comment:
I prefer
`Q.parallel_for(sycl::range<1>(n), [=](sycl::id<1> i) { b[i] = fn(a[i]); }).wait()`
In theory, it's also more efficient.